### PR TITLE
Faster methSeg

### PR DESCRIPTION
**Pull request by katwre**
Hi, 

Here I added an argument `estimate.params.density` to methSeg function to speed it up by first sampling data and to estimate on it parameters for mclust::densityMclust() (parameters: G=number of segment classes an modelName = name of a model used e.g. "V" is a variable variance (one-dimensional) ) and then run it on whole dataset. `estimate.params.density` indicates a percentage of data to sample.
Maybe you can find it useful

Kasia

### CODE REVIEW

**Comment by alexg9010** [comment]
Nice! Thank you @katwre !!

**Comment by al2na** [comment]
please test this when user try to pass named arguments to densityMclust, i would not introduce a new argument, just show it in an example how to do it. It really doesn't require that much code to make it work, you just have to pass arguments intended for densityMclust to methSeg() function

**Comment by alexg9010** [comment]
There is one problem if the user wants to have subsampling with parameters from densityMclust:
You need to know the number of segments used for the clustering in advance.
```R
## If I know the number of segments, I can easily subsample
methSeg(methylDiff.obj,diagnostic.plot = TRUE,
verbose=FALSE,
initialization=list(subset = sample(1:13,# length(seg.res)
                                                          9,
                                                          replace = FALSE)))

```

File: R/methSeg.R [hunk 1]
```diff
@@ -133,27 +142,42 @@ methSeg<-function(obj, diagnostic.plot=TRUE, join.neighbours=FALSE,  ...){
...
+    
+    nbr.sample = floor(length(seg.res) * estimate.params.density)
+    # estimate parameters for mclust
+    args.Mclust.esti = args.Mclust
+    args.Mclust.esti[["score.gr"]]=seg.res[ sample(1:length(seg.res), nbr.sample) ]
+    args.Mclust.esti[["diagnostic.plot"]]=FALSE
+    dens_estimate=do.call("densityFind", args.Mclust.esti)
+    
+    args.Mclust[["modelName"]] = dens_estimate$modelName
+    args.Mclust[["G"]] = 1:dens_estimate$G
```
Comment (al2na) [inline R/methSeg.R]: this doesn't help, you run the model twice with less parameters in the end, but this wasn't what I had in mind to make things faster

**Comment by al2na** [comment]
please use "initialization" argument, if anything. What you are doing is inefficient. 
The code below uses first 1000 segments to initialize parameters for mixture modeling, if this is not done, it will use the whole dataset for clustering and then finding initial parameters. What you have done just decreases the number of parameters and still will be slow if you have a lot of segments. you can use the similar argument to sample random segments and use it within the "initialization" argument.

```
res=methSeg(mbw,minSeg=10,G=1:4,
            initialization=list(subset=1:1000),
            join.neighbours = TRUE)
```

**Comment by al2na** [comment]
also check here:
https://stackoverflow.com/questions/14337397/how-to-run-mclust-faster-on-50000-records-dataset

**Comment by alexg9010** [comment]
Yes, I already saw that post. You are right, we actually only need give it some indices or logicals. 
My only problem is now, that when I tried this on our very small example data,  mclust crashed again for too small samples , so we need to catch these cases before. For instance I was using this code:
```
# this worked
  methSeg(methylDiff.obj,diagnostic.plot = TRUE,verbose=FALSE,initialization=list(subset = sample(1:13,# length(seg.res)
                                                                                                  9,
                                                                                                  replace = FALSE)))

# this failed
  methSeg(methylDiff.obj,diagnostic.plot = TRUE,verbose=FALSE,initialization=list(subset = sample(1:13,# length(seg.res)
                                                                                                  8,
                                                                                                  replace = FALSE)))

```

So I need to catch these cases.

**Comment by al2na** [comment]
but these are extreme cases, also in your code you need to have seg.res somewhere, seg.res is created within methSeg() to be meaningful. I think you guys can use a new argument called "initialize.on.subset" can be a number between 0 and 1, or if larger than 1 it should be taken as number of samples, then just sample rows from output of fastseg() and use that in initialization=list(subset=...) . but please check what happens when the user also provides "initialization" argument, in that case we should not run our initialization but users argument should over-ride. This is really quite easy code.

**Pull request closed without merging**